### PR TITLE
Enable denoiser for rpr20

### PR DESCRIPTION
**Pull request by DMotorygin**
JIRA TICKET: RPRMAYA-2447

PURPOSE: Denoiser working for RPR20

EFFECT FOR THE USER: Denosier denoising image when RPR20 is selected, denoising happens after redner is complete (like in Blender), denoiser working in IPR

### CODE REVIEW

File: FireRender.Maya.Src/Context/FireRenderContext.cpp [hunk 1]
```diff
@@ -424,25 +424,30 @@ bool FireRenderContext::buildScene(bool isViewport, bool glViewport, bool freshe
 	return true;
 }
 
+static const std::vector<int> denoiser_aovs = {
```
Comment (Vsevolod1983) [inline FireRender.Maya.Src/Context/FireRenderContext.cpp:427]: 1. Please rename variable if tis intended to be global
2. Should we move it to class scope ?

File: FireRender.Maya.Src/Context/FireRenderContext.cpp [hunk 2]
```diff
@@ -424,25 +424,30 @@ bool FireRenderContext::buildScene(bool isViewport, bool glViewport, bool freshe
...
 {
-	static const std::vector<int> aovsToAdd = { RPR_AOV_SHADING_NORMAL, RPR_AOV_WORLD_COORDINATE,
-		RPR_AOV_OBJECT_ID, RPR_AOV_DEPTH, RPR_AOV_DIFFUSE_ALBEDO };
-
 	// Turn on necessary AOVs
-	forceTurnOnAOVs(aovsToAdd, allocBuffer);	
+	forceTurnOnAOVs(denoiser_aovs, allocBuffer);
 }
 
+static const std::vector<int> contour_aovs = {
```
Comment (Vsevolod1983) [inline FireRender.Maya.Src/Context/FireRenderContext.cpp:441]: rename global variable to not looking as local one

File: FireRender.Maya.Src/Context/FireRenderContext.cpp [hunk 3]
```diff
@@ -1416,10 +1421,12 @@ void FireRenderContext::DebugDumpAOV(int aov) const
 	};
 
 	std::stringstream ssFileNameResolved;
+	ssFileNameResolved << "C:/temp/dbg/";
```
Comment (Vsevolod1983) [inline FireRender.Maya.Src/Context/FireRenderContext.cpp:1424]: ???

File: FireRender.Maya.Src/Context/FireRenderContext.cpp [hunk 4]
```diff
@@ -1416,10 +1421,12 @@ void FireRenderContext::DebugDumpAOV(int aov) const
...
 	ssFileNameResolved << aovNames[aov] << "_resolved.png";
 	rprFrameBufferSaveToFile(m.framebufferAOV_resolved[aov].Handle(), ssFileNameResolved.str().c_str());
 
 	std::stringstream ssFileNameNOTResolved;
+	ssFileNameNOTResolved << "C:/temp/dbg/";
```
Comment (Vsevolod1983) [inline FireRender.Maya.Src/Context/FireRenderContext.cpp:1429]: ???

File: FireRender.Maya.Src/Context/FireRenderContext.cpp [hunk 5]
```diff
@@ -1416,10 +1421,12 @@ void FireRenderContext::DebugDumpAOV(int aov) const
...
 	ssFileNameNOTResolved << aovNames[aov] << "_NOTresolved.png";
```
Comment (Vsevolod1983) [inline FireRender.Maya.Src/Context/FireRenderContext.cpp:1430]: what is it for ? debug output ?

File: FireRender.Maya.Src/FireRenderAOV.cpp [hunk 6]
```diff
@@ -20,6 +20,9 @@ limitations under the License.
 #include <maya/MViewport2Renderer.h>
 #include <maya/MGlobal.h>
 
+#include <thread>
```
Comment (Vsevolod1983) [inline FireRender.Maya.Src/FireRenderAOV.cpp:23]: is it necessary ?

File: FireRender.Maya.Src/FireRenderMaterialSwatchRender.cpp [hunk 7]
```diff
@@ -146,7 +146,7 @@ void FireRenderMaterialSwatchRender::processFromBackgroundThread()
 			(swatchInstance.getContext().height() != (unsigned int) m_resolution))
 		{
 			swatchInstance.getContext().setResolution(m_resolution, m_resolution, false);
-			swatchInstance.getContext().ConsiderSetupDenoiser();
+			swatchInstance.getContext().TryCreateDenoiserImageFilters();
```
Comment (Vsevolod1983) [inline FireRender.Maya.Src/FireRenderMaterialSwatchRender.cpp:149]: Please Remove. We don't use denoiser for swatches!

File: FireRender.Maya.Src/FireRenderProduction.cpp [hunk 8]
```diff
@@ -249,163 +255,180 @@ bool FireRenderProduction::start()
...
 	MString renderStamp;
 	if (m_globals.useRenderStamp)
 		renderStamp = m_globals.renderStampText;
 
 	m_aovs = &m_globals.aovs;
 
-	int contextWidth = m_width;
-	int contextHeight = m_height;
+	{
+		AutoMutexLock contextCreationLock(m_contextCreationLock);
```
Comment (Vsevolod1983) [inline FireRender.Maya.Src/FireRenderProduction.cpp:265]: why we need this mutex on context creation phase ?

File: FireRender.Maya.Src/FireRenderProduction.cpp [hunk 9]
```diff
@@ -249,163 +255,180 @@ bool FireRenderProduction::start()
...
+	return true;
 }
 
-void FireRenderProduction::OnBufferAvailableCallback()
+void FireRenderProduction::OnBufferAvailableCallback(float progress)
 {
 	AutoMutexLock pixelsLock(m_pixelsLock);
-	m_renderViewAOV->readFrameBuffer(*m_contextPtr, true);
+
+	bool frameFinished = fabs(1.0f - progress) <= FLT_EPSILON;
```
Comment (Vsevolod1983) [inline FireRender.Maya.Src/FireRenderProduction.cpp:425]: hmm not sure about such check

**Review by Vsevolod1983 — CHANGES_REQUESTED** [review]

File: FireRender.Maya.Src/FireRenderProduction.cpp (hunk 8 above)
Comment (DMotorygin) [reply FireRender.Maya.Src/FireRenderProduction.cpp:265]: this is from legacy code

**Review by Vsevolod1983 — APPROVED** [review]

File: FireRender.Maya.Src/Context/FireRenderContext.cpp (hunk 3 above)
Comment (DMotorygin) [reply FireRender.Maya.Src/Context/FireRenderContext.cpp]: ebug output

File: FireRender.Maya.Src/FireRenderProduction.cpp (hunk 9 above)
Comment (DMotorygin) [reply FireRender.Maya.Src/FireRenderProduction.cpp:424]: richardgeslot  7:57 PM
Hi Dmitrii, yes 0.0 to 1.0
7:58
even for multiple iterations :
rprContextSetParameterByKey1u(context_, RPR_CONTEXT_ITERATIONS, N);
this callback will be 0.0 to 1.0

**Pull request merged**